File: projects/07/NameTester.rb
# NameTester.rb tests class Name and its operations
#
# Begun by: Dr. Adams, for CS 214 at Calvin College.
# Completed by: Jacob Brink
# Date: 4/8/2020
# Project: 7
####################################################

require 'test/unit/assertions'   # needed for assert
include Test::Unit::Assertions

class Name

  def initialize(first, middle, last)
    @first, @middle, @last = first, middle, last
  end

  attr_accessor :first, :middle, :last
  
  def fullName()
    @first + " " + @middle + " " + @last
  end

  def lfmi()
    @last + ", " + @first + " " + @middle[0] + "."
  end

  def read()
    puts "First: "
    @first = gets.chomp

    puts "Middle: "
    @middle = gets.chomp

    puts "Last: "
    @last = gets.chomp
  end

  def print()
    puts self.fullName
    self.fullName
  end
  
end


def testName
   name = Name.new("John", "Paul", "Jones")

   puts "Testing Lab Methods..."
   
   assert name.first == "John", "first failed"
   assert name.middle == "Paul", "middle failed"
   assert name.last == "Jones", "last failed"
   assert name.fullName == "John Paul Jones", "fullName failed"
   assert name.print == "John Paul Jones", "print failed"

   puts "Testing Project Methods..."
   name.first = "Jacob"
   name.middle = "Jingle"
   name.last = "Hymer"

   assert name.first == "Jacob", "first setter failed"
   assert name.middle == "Jingle", "middle setter failed"
   assert name.last == "Hymer", "last setter failed"
   assert name.lfmi == "Hymer, Jacob J.", "lfmi failed"

   name.read()
   puts "Does this look right?"
   name.print
   
   print "All tests passed!\n"
end

if __FILE__ == $0
  testName()
end

File: projects/07/NameTester.java
/* NameTester.java is a 'driver' to test class Name.
 * Begun by: Prof. Adams, for CS 214 at Calvin College.
 * Student: Jacob Brink
 * Date: 3/20/2020
 * Project: 07
 ************************************************************/

import java.io.*;
import java.util.Scanner;


public class NameTester  
{


  private static void runProjectTests()
  {
      Name aName = new Name("John", "Paul", "Jones");

      //test mutator methods
      System.out.println("Testing mutator methods");
      
      aName.setFirst("First");
      assert aName.getFirst().equals("First");
      assert aName.getMiddle().equals("Paul");
      assert aName.getLast().equals("Jones");
      
      aName.setMiddle("Middle");
      assert aName.getFirst().equals("First");
      assert aName.getMiddle().equals("Middle");
      assert aName.getLast().equals("Jones");

      
      aName.setLast("Last");
      assert aName.getFirst().equals("First");
      assert aName.getMiddle().equals("Middle");
      assert aName.getLast().equals("Last");

      //testing lfmi method
      System.out.println("Testing lfmi method");
      Name bName = new Name("Jacob", "Hymer", "Smith");
      assert bName.lfmi().equals("Smith, Jacob H.");

      //testing input method
      System.out.println("Testing read method");
      InputStream emptyStream = System.in;
      //reset input stream in case user types during test
      System.setIn(emptyStream);
      ByteArrayInputStream in = new ByteArrayInputStream(("Mike" + System.lineSeparator() + "Peter" + System.lineSeparator() + "Smith").getBytes());
      System.setIn(in);
      aName.read(new Scanner(System.in));
      assert aName.getFirst().equals("Mike");
      assert aName.getMiddle().equals("Peter");
      assert aName.getLast().equals("Smith");
      

  }
    
  public static void main(String[] args)
  {
      
      Name aName = new Name("John", "Paul", "Jones");
//
      assert aName.getFirst().equals("John") ;
      assert aName.getMiddle().equals("Paul") ;
      assert aName.getLast().equals("Jones") ;
//
      System.out.println(aName);
      assert aName.toString().equals("John Paul Jones");

      runProjectTests();
      
      System.out.println("All tests passed!");
  }
}

class Name
{
    private String myFirst, myMiddle, myLast;


    public Name(String first, String middle, String last)
    {
	
	myFirst = first;
	myMiddle = middle;
	myLast = last;
    }


    public void read(Scanner scanner) {
	myFirst = scanner.nextLine();
	myMiddle = scanner.nextLine();
	myLast = scanner.nextLine();
    }
    
    public String lfmi()
    {
	return String.format("%s, %s %s.", myLast, myFirst, myMiddle.charAt(0));
    }

    public void setFirst(String first)
    {
	this.myFirst = first;
    }

    public void setMiddle(String middle)
    {
	this.myMiddle = middle;
    }

    public void setLast(String last)
    {
	this.myLast = last;
    }

    public String getFirst()
    {
	return this.myFirst;
    }

    public String getMiddle()
    {
	return this.myMiddle;
    }

    public String getLast()
    {
	return this.myLast;
    }

    public String toString()
    {
	return myFirst + ' ' + myMiddle + ' ' + myLast;
    }
}


